ext/http/http_next.rs

```diff
@@ -0,0 +1,800 @@
+type Response = hyper1::Response<ResponseBytes>;
+
+pub struct HttpSlabRecord {
+  request_info: HttpConnectionProperties,
+  request_parts: Parts,
+  request_body: Option<Incoming>,
+  // The response may get taken before we tear this down
+  response: Option<Response>,
+  body: Option<Rc<HttpRequestBody>>,
+  promise: CompletionHandle,
+  #[cfg(zombie_http_tracking)]
+  alive: bool,
+}
+
+thread_local! {
+  pub static SLAB: RefCell<Slab<HttpSlabRecord>> = RefCell::new(Slab::with_capacity(1024));
+}
+
+macro_rules! with {
+  ($ref:ident, $mut:ident, $type:ty, $http:ident, $expr:expr) => {
+    #[inline(always)]
+    #[allow(dead_code)]
+    pub(crate) fn $mut<T>(key: usize, f: impl FnOnce(&mut $type) -> T) -> T {
+      SLAB.with(|slab| {
+        let mut borrow = slab.borrow_mut();
+        #[allow(unused_mut)] // TODO(mmastrac): compiler issue?
+        let mut $http = borrow.get_mut(key).unwrap();
+        #[cfg(zombie_http_tracking)]
+        if !$http.alive {
+          panic!("Attempted to access a dead HTTP object")
+        }
+        f(&mut $expr)
+      })
+    }
+
+    #[inline(always)]
+    #[allow(dead_code)]
+    pub(crate) fn $ref<T>(key: usize, f: impl FnOnce(&$type) -> T) -> T {
+      SLAB.with(|slab| {
+        let borrow = slab.borrow();
+        let $http = borrow.get(key).unwrap();
+        #[cfg(zombie_http_tracking)]
+        if !$http.alive {
+          panic!("Attempted to access a dead HTTP object")
+        }
+        f(&$expr)
+      })
+    }
+  };
+}
+
+with!(with_req, with_req_mut, Parts, http, http.request_parts);
+with!(
+  with_req_body,
+  with_req_body_mut,
+  Option<Incoming>,
+  http,
+  http.request_body
+);
+with!(
+  with_resp,
+  with_resp_mut,
+  Option<Response>,
+  http,
+  http.response
+);
+with!(
+  with_body,
+  with_body_mut,
+  Option<Rc<HttpRequestBody>>,
+  http,
+  http.body
+);
+with!(
+  with_promise,
+  with_promise_mut,
+  CompletionHandle,
+  http,
+  http.promise
+);
+with!(with_http, with_http_mut, HttpSlabRecord, http, http);
+
+fn slab_insert(
+  request: Request,
+  request_info: HttpConnectionProperties,
+) -> usize {
+  SLAB.with(|slab| {
+    let (request_parts, request_body) = request.into_parts();
+    slab.borrow_mut().insert(HttpSlabRecord {
+      request_info,
+      request_parts,
+      request_body: Some(request_body),
+      response: Some(Response::new(ResponseBytes::default())),
+      body: None,
+      promise: CompletionHandle::default(),
+      #[cfg(zombie_http_tracking)]
+      alive: true,
+    })
+  })
+}
+
+struct SafeFutureForSingleThread<F: Future<Output = O> + Unpin, O>(F);
+
+impl<F: Future<Output = O> + Unpin, O> Future
+  for SafeFutureForSingleThread<F, O>
+{
+  type Output = O;
+  fn poll(
+    mut self: Pin<&mut Self>,
+    cx: &mut std::task::Context<'_>,
+  ) -> std::task::Poll<Self::Output> {
+    self.0.poll_unpin(cx)
+  }
+}
+
+// SAFETY: We use this to allow us to use tokio::task::spawn in a single-threaded executor.
```

Comment:
Could you explain why it's fine to do so?

Reply:
Removed this code as we no longer need it.